buildSrc/src/main/java/org/elasticsearch/gradle/precommit/PrecommitPlugin.java

```diff
@@ -72,7 +72,18 @@ public void apply(Project project) {
                     p -> project.getTasks().named(LifecycleBasePlugin.CHECK_TASK_NAME).configure(t -> t.dependsOn(precommit))
                 );
             project.getPluginManager()
-                .withPlugin("java", p -> project.getTasks().withType(Test.class).configureEach(t -> t.mustRunAfter(precommit)));
+                .withPlugin(
```

Comment:
I still don't quite understand the need for the separate `PrecommitTaskPlugin`. How does it differ from the `PrecommitPlugin` which also applies it? Couldn't we must move that logic up into the `PrecommitPlugin` and use that across the board?

Reply:
Oh I see, I'm getting confused by the fact that it's an inner class, also the fact that the abstract `PrecommitPlugin` actually does include some logic order java compilation tasks.

I'm still not sure the java plugin logic should live in the `PrecommitTaskPlugin`. If it's purpose is just to create the precommit task then we should limit it there and move that logic to the `ElasticsearchJavaPlugin`. I'd also extract it as it's own class instead of being an inner class if we expect it to be used by other plugins. I'd probably also rename it to `ElasticsearchLifecyclePlugin` to comply with the similar convention used in Gradle, and we can add any other lifecycle tasks we create to it in the future.

Reply:
The existing java plugin ordering is, I think, not possible outside of the current implementation. We make sure that compilation happens before the rest of the precommit tasks. We do that by adding a shoudlRunAfter when the task of each PrecommitPlugin is applied. We would have to iterate over all the dependencies of precommit, instead of applying to each task as it is added, which I'm not sure it possible?

Reply:
Yeah, the ordering stuff would be tricky to strip out, that can likely stay. The only way to do it "properly" would be to have all precommit tasks implement some kind of marker interface, then have the Java plugin add the ordering rule but we often create tasks that we don't own the implementation of, so that's not practically feasible.

Reply:
I moved the plugin out to its own file in 06016fc